### ISSUE TITLE
hpx +rocm build fails: unity_0_cxx.cxx.o: boost/container/vector.hpp:2740:4: error: use of undeclared identifier 'noinline'; did you mean 'inline'?

### ISSUE THREAD

**Issue by eugeneswalker** [LLNL/spack#27875]
### Steps to reproduce the issue

`hpx@1.7.1 +rocm amdgpu_target=gfx908` fails to build using:
* `spack@develop` (f81d84dfc6b1c3a821dffab785e62ecd79f130ab from `Tue Dec 7 12:30:14 2021 +0100`
* Ubuntu 20.04, x86_64
* GCC 9.3.0
* ROCm 4.3.1, defined externally, installed via apt

Concrete spec: [hpx.spec.json.txt](https://github.com/spack/spack/files/7681217/hpx.spec.json.txt)

Concretization:
```
hpx@1.7.1%gcc@9.3.0~async_cuda~async_mpi~cuda~examples~generic_coroutines~ipo+rocm~tools amdgpu_target=gfx908 build_type=RelWithDebInfo cxxstd=17 instrumentation=none malloc=tcmalloc max_cpu_count=64 networking=tcp arch=linux-ubuntu20.04-x86_64
    ^asio@1.20.0%gcc@9.3.0~boost_coroutine~boost_regex~separate_compilation cxxstd=17 arch=linux-ubuntu20.04-x86_64
        ^autoconf@2.69%gcc@9.3.0 patches=35c449281546376449766f92d49fc121ca50e330e60fefcfc9be2af3253082c2,7793209b33013dc0f81208718c68440c5aae80e7a1c4b8d336e382525af791a7,a49dd5bac3b62daa0ff688ab4d508d71dbd2f4f8d7e2a02321926346161bf3ee arch=linux-ubuntu20.04-x86_64
            ^m4@1.4.19%gcc@9.3.0+sigsegv patches=9dc5fbd0d5cb1037ab1e6d0ecc74a30df218d0a94bdd5a02759a97f62daca573,bfdffa7c2eb01021d5849b36972c069693654ad826c1a20b53534009a4ec7a89 arch=linux-ubuntu20.04-x86_64
                ^libsigsegv@2.13%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^perl@5.34.0%gcc@9.3.0+cpanm+shared+threads arch=linux-ubuntu20.04-x86_64
                ^berkeley-db@18.1.40%gcc@9.3.0+cxx~docs+stl patches=b231fcc4d5cff05e5c3a4814f6a5af0e9a966428dc2176540d2c05aff41de522 arch=linux-ubuntu20.04-x86_64
                ^bzip2@1.0.8%gcc@9.3.0~debug~pic+shared arch=linux-ubuntu20.04-x86_64
                    ^diffutils@3.8%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^libiconv@1.16%gcc@9.3.0 libs=shared,static arch=linux-ubuntu20.04-x86_64
                ^gdbm@1.19%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^readline@8.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^ncurses@6.2%gcc@9.3.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-x86_64
                            ^pkgconf@1.8.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^zlib@1.2.11%gcc@9.3.0+optimize+pic+shared arch=linux-ubuntu20.04-x86_64
        ^automake@1.16.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^boost@1.77.0%gcc@9.3.0+atomic+chrono~clanglibcpp~container~context~coroutine+date_time~debug+exception~fiber+filesystem+graph~icu+iostreams+locale+log+math~mpi+multithreaded~numpy~pic+program_options~python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer~versionedlayout+wave cxxstd=17 patches=93f4aad8f88d1437e50d95a2d066390ef3753b99ef5de24f7a46bc083bd6df06 visibility=global arch=linux-ubuntu20.04-x86_64
        ^libtool@2.4.6%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
    ^cmake@3.21.4%gcc@9.3.0~doc+ncurses+openssl+ownlibs~qt build_type=Release arch=linux-ubuntu20.04-x86_64
        ^openssl@1.1.1l%gcc@9.3.0~docs certs=system arch=linux-ubuntu20.04-x86_64
    ^git@2.31.1%gcc@9.3.0+man+nls+perl~svn~tcltk arch=linux-ubuntu20.04-x86_64
        ^curl@7.80.0%gcc@9.3.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2~nghttp2 tls=openssl arch=linux-ubuntu20.04-x86_64
        ^expat@2.4.1%gcc@9.3.0+libbsd arch=linux-ubuntu20.04-x86_64
            ^libbsd@0.11.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^libmd@1.0.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^gettext@0.21%gcc@9.3.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu20.04-x86_64
            ^libxml2@2.9.12%gcc@9.3.0~python arch=linux-ubuntu20.04-x86_64
                ^xz@5.2.5%gcc@9.3.0+pic libs=shared,static arch=linux-ubuntu20.04-x86_64
            ^tar@1.34%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^libidn2@2.3.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^libunistring@0.9.10%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^openssh@8.7p1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^libedit@3.1-20210216%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^pcre2@10.36%gcc@9.3.0~jit+multibyte arch=linux-ubuntu20.04-x86_64
    ^gperftools@2.9.1%gcc@9.3.0+debugalloc~dynamic_sized_delete_support+libunwind~sized_delete arch=linux-ubuntu20.04-x86_64
        ^libunwind@1.5.0%gcc@9.3.0~block_signals~conservative_checks~cxx_exceptions~debug~debug_frame+docs+pic+tests+weak_backtrace+xz~zlib components=none libs=shared,static arch=linux-ubuntu20.04-x86_64
    ^hip@4.3.1%gcc@9.3.0~ipo build_type=Release patches=2a4190477b7d9206b9cd8d70770ba0bc007273cbe54772efb12f9ca2e37c0392,99190b4616edb362d48f9b265c3018a3c6339481b0729d9fe46185fca25bc54b,e276c4acf3d37712b6bea306fea34f539d3c4f743471e9da208b5eb17b16ae67 arch=linux-ubuntu20.04-x86_64
    ^hsa-rocr-dev@4.3.1%gcc@9.3.0+image~ipo+shared build_type=Release patches=71e6851d9be8113bfb8d71b66a3171e0d0401bae5e6f161c9e7fe32558261a46 arch=linux-ubuntu20.04-x86_64
    ^hwloc@2.6.0%gcc@9.3.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~opencl+pci~rocm+shared arch=linux-ubuntu20.04-x86_64
        ^libpciaccess@0.16%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^util-macros@1.19.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
    ^llvm-amdgpu@4.3.1%gcc@9.3.0~ipo+openmp+rocm-device-libs build_type=Release patches=d999f3b235e655ee07f6dd2590302082feaa06d32c5c6b53aae9c5cf1e45b644 arch=linux-ubuntu20.04-x86_64
    ^ninja@1.10.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^python@3.8.12%gcc@9.3.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87,4c2457325f2b608b1b6a2c63087df8c26e07db3e3d493caf36a56f0ecf6fb768,f2fd060afc4b4618fe8104c4c5d771f36dc55b1db5a4623785a4ea707ec72fb4 arch=linux-ubuntu20.04-x86_64
            ^libffi@3.3%gcc@9.3.0 patches=26f26c6f29a7ce9bf370ad3ab2610f99365b4bdd7b82e7c31df41a3370d685c0 arch=linux-ubuntu20.04-x86_64
            ^sqlite@3.36.0%gcc@9.3.0+column_metadata+fts~functions~rtree arch=linux-ubuntu20.04-x86_64
            ^util-linux-uuid@2.36.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
```

```
$> spack install -f ./hpx.spec.json
...
==> Installing hpx-1.7.1-xahhyffysdiox4kq27eozgol7rvdrk5q
==> No binary for hpx-1.7.1-xahhyffysdiox4kq27eozgol7rvdrk5q found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/00/008a0335def3c551cba31452eda035d7e914e3e4f77eec679eea070ac71bd83b.tar.gz
==> No patches needed for hpx
==> hpx: Executing phase: 'cmake'
...
  >> 636    FAILED: plugins/parcelport/tcp/CMakeFiles/parcelport_tcp.dir/Unity/unity_0_cxx.cxx.o
...
     638    In file included from /tmp/root/spack-stage/spack-stage-hpx-1.7.1-xahhyffysdiox4kq27eozgol7rvdrk5q/spack-build-xahhyff/plugins/parcelport/tcp/CMakeFiles/parcelport_tcp.
            dir/Unity/unity_0_cxx.cxx:3:
     639    In file included from /tmp/root/spack-stage/spack-stage-hpx-1.7.1-xahhyffysdiox4kq27eozgol7rvdrk5q/spack-src/plugins/parcelport/tcp/connection_handler_tcp.cpp:18:
     640    In file included from /tmp/root/spack-stage/spack-stage-hpx-1.7.1-xahhyffysdiox4kq27eozgol7rvdrk5q/spack-src/libs/parallelism/futures/include/hpx/futures/future.hpp:19:
     641    In file included from /tmp/root/spack-stage/spack-stage-hpx-1.7.1-xahhyffysdiox4kq27eozgol7rvdrk5q/spack-src/libs/parallelism/futures/include/hpx/futures/detail/future_
            data.hpp:28:
     642    In file included from /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/boost-1.77.0-ogvzijww2qp4nvexipvlsfpqfmzsytmt/include/boost/container/small_vector.hpp:27:
  >> 643    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/boost-1.77.0-ogvzijww2qp4nvexipvlsfpqfmzsytmt/include/boost/container/vector.hpp:2740:4: error: use of undeclared id
            entifier 'noinline'; did you mean 'inline'?
     644       BOOST_CONTAINER_NOINLINE iterator priv_insert_forward_range_no_capacity
     645       ^
     646    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/boost-1.77.0-ogvzijww2qp4nvexipvlsfpqfmzsytmt/include/boost/container/detail/workaround.hpp:116:37: note: expanded f
            rom macro 'BOOST_CONTAINER_NOINLINE'
     647       #define BOOST_CONTAINER_NOINLINE BOOST_NOINLINE
     648                                        ^
     649    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/boost-1.77.0-ogvzijww2qp4nvexipvlsfpqfmzsytmt/include/boost/config/detail/suffix.hpp:633:46: note: expanded from mac
            ro 'BOOST_NOINLINE'

     ...

     654                                        ^
     655    In file included from /tmp/root/spack-stage/spack-stage-hpx-1.7.1-xahhyffysdiox4kq27eozgol7rvdrk5q/spack-build-xahhyff/plugins/parcelport/tcp/CMakeFiles/parcelport_tcp.
            dir/Unity/unity_0_cxx.cxx:3:
     656    In file included from /tmp/root/spack-stage/spack-stage-hpx-1.7.1-xahhyffysdiox4kq27eozgol7rvdrk5q/spack-src/plugins/parcelport/tcp/connection_handler_tcp.cpp:18:
     657    In file included from /tmp/root/spack-stage/spack-stage-hpx-1.7.1-xahhyffysdiox4kq27eozgol7rvdrk5q/spack-src/libs/parallelism/futures/include/hpx/futures/future.hpp:19:
     658    In file included from /tmp/root/spack-stage/spack-stage-hpx-1.7.1-xahhyffysdiox4kq27eozgol7rvdrk5q/spack-src/libs/parallelism/futures/include/hpx/futures/detail/future_
            data.hpp:28:
     659    In file included from /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/boost-1.77.0-ogvzijww2qp4nvexipvlsfpqfmzsytmt/include/boost/container/small_vector.hpp:27:
  >> 660    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/boost-1.77.0-ogvzijww2qp4nvexipvlsfpqfmzsytmt/include/boost/container/vector.hpp:2740:4: error: type name does not a
            llow function specifier to be specified
     661    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/boost-1.77.0-ogvzijww2qp4nvexipvlsfpqfmzsytmt/include/boost/container/detail/workaround.hpp:116:37: note: expanded f
            rom macro 'BOOST_CONTAINER_NOINLINE'
     662       #define BOOST_CONTAINER_NOINLINE BOOST_NOINLINE
     663                                        ^
     664    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/boost-1.77.0-ogvzijww2qp4nvexipvlsfpqfmzsytmt/include/boost/config/detail/suffix.hpp:633:46: note: expanded from mac
            ro 'BOOST_NOINLINE'
     665    #      define BOOST_NOINLINE __attribute__ ((__noinline__))
     666                                                 ^

     ...

     669                                        ^
     670    In file included from /tmp/root/spack-stage/spack-stage-hpx-1.7.1-xahhyffysdiox4kq27eozgol7rvdrk5q/spack-build-xahhyff/plugins/parcelport/tcp/CMakeFiles/parcelport_tcp.
            dir/Unity/unity_0_cxx.cxx:3:
     671    In file included from /tmp/root/spack-stage/spack-stage-hpx-1.7.1-xahhyffysdiox4kq27eozgol7rvdrk5q/spack-src/plugins/parcelport/tcp/connection_handler_tcp.cpp:18:
     672    In file included from /tmp/root/spack-stage/spack-stage-hpx-1.7.1-xahhyffysdiox4kq27eozgol7rvdrk5q/spack-src/libs/parallelism/futures/include/hpx/futures/future.hpp:19:
     673    In file included from /tmp/root/spack-stage/spack-stage-hpx-1.7.1-xahhyffysdiox4kq27eozgol7rvdrk5q/spack-src/libs/parallelism/futures/include/hpx/futures/detail/future_
            data.hpp:28:
     674    In file included from /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/boost-1.77.0-ogvzijww2qp4nvexipvlsfpqfmzsytmt/include/boost/container/small_vector.hpp:27:
  >> 675    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/boost-1.77.0-ogvzijww2qp4nvexipvlsfpqfmzsytmt/include/boost/container/vector.hpp:2740:4: error: expected expression
     676    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/boost-1.77.0-ogvzijww2qp4nvexipvlsfpqfmzsytmt/include/boost/container/detail/workaround.hpp:116:37: note: expanded f
            rom macro 'BOOST_CONTAINER_NOINLINE'
     677       #define BOOST_CONTAINER_NOINLINE BOOST_NOINLINE
     678                                        ^
     679    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/boost-1.77.0-ogvzijww2qp4nvexipvlsfpqfmzsytmt/include/boost/config/detail/suffix.hpp:633:46: note: expanded from mac
            ro 'BOOST_NOINLINE'
     680    #      define BOOST_NOINLINE __attribute__ ((__noinline__))
     681
...
  >> 734    FAILED: plugins/parcel/coalescing/CMakeFiles/parcel_coalescing.dir/Unity/unity_0_cxx.cxx.o
...
```

### Information on your system

* **Spack:** 0.17.0-446-3db39db0b3
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-broadwell
* **Concretizer:** clingo

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/7681200/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/7681202/spack-build-env.txt)
[spack-build-01-cmake-out.txt](https://github.com/spack/spack/files/7681209/spack-build-01-cmake-out.txt)

@albestro 
@msimberg 
@teonnik 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by msimberg** [comment]
Thanks @eugeneswalker. This is a known issue with HIP and Boost: https://github.com/boostorg/config/issues/392. It has been fixed in the _just_ released Boost 1.78.0. I'll open a PR to add a conflict for this.

**Comment by eugeneswalker** [comment]
> Thanks @eugeneswalker. This is a known issue with HIP and Boost: [boostorg/config#392](https://github.com/boostorg/config/issues/392). It has been fixed in the _just_ released Boost 1.78.0. I'll open a PR to add a conflict for this.

Thank you!

**Comment by eugeneswalker** [comment]
Yes, can confirm on my end that `hpx +rocm` builds when built against `boost@1.78.0` -- Thank you! Closing this issue.
